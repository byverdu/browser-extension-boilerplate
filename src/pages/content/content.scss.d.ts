// *** This file is automatically generated - Do not edit! ***
/* tslint:disable */
/* eslint-disable */

declare const styles = {
    content: 'content__content--2-DP5',
} as const;

export type ClassName = string;

export default styles;