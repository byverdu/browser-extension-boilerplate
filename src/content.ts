import { wrapperBrowserAPI } from 'api';
import { Link } from 'types';

const { setStorage, sendMessage } = wrapperBrowserAPI;
const allLinksInTab: Link[] = [
  ...document.querySelectorAll ( 'a' ),
]
  .map ( ( { href, textContent } ) => ( { href, textContent } ) )
  .filter ( ( link ) => link.textContent );


const div: HTMLDivElement = document.createElement ( 'div' );

setStorage ( 'links-saved', allLinksInTab  )
  .then ( () => {
    sendMessage ( 'links-saved' )
      .then ( resp => {
        if ( typeof resp === 'boolean' ) {
          div.textContent = `Hello.... this page has ${allLinksInTab.length} links on it`;
          document.body.appendChild ( div );
        }
      } );
  } );