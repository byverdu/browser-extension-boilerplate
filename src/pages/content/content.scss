.content {
  position: absolute;
  top: 0;
  z-index: 999999;
  padding: 10px;
  background: red;
  color: white;
}
