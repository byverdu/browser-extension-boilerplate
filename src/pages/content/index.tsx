import React, { useState, useEffect, useRef } from 'react';
import { render } from 'react-dom';

import { wrapperBrowserAPI } from 'api';
import { Link,ExtensionMessages } from 'types';

import styles from './content.scss';

const { setStorage, sendMessage, getExtensionName } = wrapperBrowserAPI;
const extensionName = getExtensionName ();
const linksSaved: ExtensionMessages = 'links-saved'

const App = () => {
  const [ linksCount, setLinksCount ] = useState<number> ( 0 );
  const container = useRef<HTMLDivElement> ();
  const saveLinks = async ( links ) => {
    const oldStorage = await wrapperBrowserAPI.getStorage ( linksSaved );
    const newValues = {
      [ linksSaved ]: {
        ...oldStorage[ linksSaved ],
        ...links
      }
    };

    return await setStorage ( linksSaved, newValues );
  };

  useEffect ( () => {
    const links: Link[] = [
      ...document.querySelectorAll ( 'a' ),
    ]
      .map ( ( { href, textContent } ) => ( { href, textContent } ) )
      .filter ( ( link ) => link.textContent );

    saveLinks ( links ).then ( ( resp ) => {
      console.log ( 'storage', resp );
      sendMessage ( linksSaved )
        .then ( resp => {
          if ( typeof resp === 'boolean' ) {
            setLinksCount ( links.length );
          }
        } );
    } );
  } );

  useEffect ( () => {
    setTimeout ( () => {
      if ( container.current ) {
        container.current.remove ();
      }
    }, 5000 );
  } );

  return (
    <div ref={container} className={styles.content}>
      Hello.... this page has {linksCount} links on it
    </div>
  );
};

const root = document.createElement ( 'div' );
root.setAttribute ( 'id',  extensionName );
document.body.appendChild ( root );

render ( <App />, document.getElementById ( extensionName ) );
