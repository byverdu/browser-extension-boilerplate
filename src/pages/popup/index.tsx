import React, { useState, useEffect, useCallback } from 'react';
import { render } from 'react-dom';

import { GetActiveTabResp, Link } from 'types/extension.types';

import { wrapperBrowserAPI, MESSAGES_TYPES } from 'api/extension.api';

import { Header } from 'components/Header';
import { List } from 'components/List';

import styles from './popup.scss';

const { getStorage, sendMessage, sendTabMessage } = wrapperBrowserAPI;
const { LINKS_SAVED, FIND_LINK, GET_ACTIVE_TAB } = MESSAGES_TYPES;

interface State{
  links: Link[]
  tabHostname: string
  tabId: number
}

export const Popup = () => {
  const [ state, setState ] = useState<State> ( {
    links: [],
    tabHostname: '',
    tabId: undefined
  } );

  const onclickHandler = useCallback ( async ( href ) => {
    try {
      await sendTabMessage ( state.tabId, { 
        type: FIND_LINK,
        payload: href 
      } );
    } catch ( e ) {
      console.log ( 'state', state );
      console.log ( 'sendTabMessage', e );
    }
  }, [ state ] );

  useEffect (  () => {
    ( sendMessage ( { 
      type: GET_ACTIVE_TAB 
    } ) as Promise<GetActiveTabResp> ).then ( ( resp ) => {
      console.log ( GET_ACTIVE_TAB, resp );
      if ( typeof resp === 'object' ) {
        const { url, id: tabId } = resp;
        const tabHostname = new URL ( url ).hostname;

        getStorage ( LINKS_SAVED ).then ( resp => {
          setState ( {
            links: resp[ LINKS_SAVED ] [ tabHostname ],
            tabHostname,
            tabId
          } );
        } );
      }
    } );
  }, [] );

  const { links, tabHostname } = state;

  return (
    <div className={styles.popup}>
      <Header title={tabHostname} cssClass={styles.title} />
      <List
        links={links}
        onclick={onclickHandler}
      />
    </div>
  );
};

if ( process.env.ENV !== 'test' ) {
  render ( <Popup />, document.getElementById ( 'root' ) );
}
