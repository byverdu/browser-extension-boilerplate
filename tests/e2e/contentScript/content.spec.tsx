jest.mock ( '../../src/api' );
import React from 'react';
import { render, waitFor } from './custom.render';
import '@testing-library/jest-dom/extend-expect';
import { App } from '../../../src/pages/content';

describe ( 'Content script app', () => {
  it ( 'should not render a message if no links present in Document', async () => {
    const { getByText } = render ( <App />, {
      withLinks: false 
    } );

    await waitFor ( () => {

      expect ( getByText ( 'Hello.... this page has 0 links on it' ) ).toBeInTheDocument (); 
    } );
  } );

  it ( 'should render links message if links are present in Document', async () => {
    const { container, getByText } = render ( <App />, {
      withLinks: true 
    } );

    await waitFor ( () => {
      expect ( getByText ( 'Hello.... this page has 3 links on it' ) ).toBeInTheDocument (); 
      expect ( container.querySelectorAll ( 'a' ) ).toHaveLength ( 3 );
    } );
  } );

  it ( 'should remove links message after 5 miliseconds', async () => {
    jest.useFakeTimers ();
    const { getByText } = render ( <App />, {
      withLinks: true 
    } );

    await waitFor ( () => {
      const linksMessage = getByText ( 'Hello.... this page has 3 links on it' );
      jest.runAllTimers ();
      expect ( linksMessage ).not.toBeInTheDocument ();
    } );
  } );

  it ( 'should match the snapshot', async () => {
    const { container, getByText } = render ( <App />, {
      withLinks: true 
    } );

    await waitFor ( () => {
      expect ( container.firstChild ).toMatchSnapshot ( `<section>
        <a href="http://localhost/first-link">first link</a>
        <a href="http://localhost/second-link">second link</a>
        <a href="http://localhost/third-link">third link</a>
        <div class="content">Hello.... this page has 3 links on it</div>
      </section>` );
    } );
  } );
} );


