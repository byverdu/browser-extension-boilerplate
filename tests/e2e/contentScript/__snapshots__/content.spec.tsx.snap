// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Content script app should match the snapshot: <section>
        <a href="http://localhost/first-link">first link</a>
        <a href="http://localhost/second-link">second link</a>
        <a href="http://localhost/third-link">third link</a>
        <div class="content">Hello.... this page has 3 links on it</div>
      </section> 1`] = `
<section>
  <a
    href="http://localhost/first-link"
  >
    first link
  </a>
  <a
    href="http://localhost/second-link"
  >
    second link
  </a>
  <a
    href="http://localhost/third-link"
  >
    third link
  </a>
  <div
    class="content"
  >
    Hello.... this page has
     
    0
     
    links on it
  </div>
</section>
`;
